### ISSUE TITLE
Sync daemon mode

### ISSUE THREAD

**Issue by untitaker** [pimutils/vdirsyncer#24]
Long-running process that keeps connections open. Not sure if this would be that useful.
##

**Comment by hobarrera** [comment]
> Not sure if this would be that useful.

None-plaintext passwords, something that can't really be achieved by merely running it via cron.

**Comment by untitaker** [comment]
vdirsyncer prompts for the password, you can pipe to it from a wrapper script. Not ideal, but possible.

**Comment by hobarrera** [comment]
> vdirsyncer prompts for the password, you can pipe to it from a wrapper script. Not ideal, but possible.

Which sort of brings us back to #228. 😁 (I don't want to have to *interact* with anything when running via cron).

**Comment by untitaker** [comment]
>(I don't want to have to interact with anything when running via cron).

Right, but you can write a script like this:

    local password="$(pass vdirsyncer)"
    while true; do
        echo "$password" | vdirsyncer sync
        sleep 60
    done

and run it as daemon. You now have loaded the password in memory, don't have to store it as plaintext on disk and vdirsyncer still authenticates properly.

**Comment by untitaker** [comment]
And yes, that still requires you to enter your master password once, but there's no way around that.

**Comment by untitaker** [comment]
It seems like this feature is inevitable for security, but I wonder if there is another user-friendly way that doesn't involve re-inventing cronjobs in Python.

BTW, if you have installed [click-repl](https://github.com/click-contrib/click-repl), you can pipe commands to `vdirsyncer repl` and it will keep the passwords in memory.

**Comment by hobarrera** [comment]
A second upside on a daemon is that it can use something like libevent-python to *instantly* push locally-modified events (without re-scanning everything else), and only download an defined intervals.

**Comment by untitaker** [comment]
>*instantly* push locally-modified events (without re-scanning everything else)

Push events probably induce an insane amount of complexity, so I'm not sure if
I'm going to care about implementing it.  I am still encountering bugs in
offlineimap's IDLE support which is why I've turned it off.

**Comment by untitaker** [comment]
FWIW I'm currently using this: https://github.com/untitaker/dotfiles/blob/master/home/.scripts/vsy

It requires [click-repl](https://github.com/click-contrib/click-repl).

**Comment by untitaker** [comment]
I've set the milestone to 1.0. I'm currently using the workaround described above, so I have no personal incentive to actually fix this. But we definetly need a better (official) solution here rather than the combo of an undocumented feature, an unstable third-party package and a massive shell script.

**Comment by untitaker** [comment]
The script I use above no longer requires an undocumented hack. So it is possible develop a vdirsyncer daemon outside of vdirsyncer itself, and I think this would be a good idea to take a shortcut to stable 1.0.

**Comment by Decipher** [comment]
Adding my interest in getting a Daemon mode added, I'm currently in the process of putting together a Docker version of vdirsyncer (https://hub.docker.com/r/deciphered/vdirsyncer-docker/) so that I can run it as a service on my Synology NAS, and having a daemon is essential.

**Comment by untitaker** [comment]
@Decipher depending on what you would exactly expect from such a daemon a cronjob or similar might do the trick

**Comment by Decipher** [comment]
@untitaker That's what I intend on implementing to start with, yes, but the benefit of having a Daemon mode would be that the logic to determine if the previous sync is still running or how long since the last sync was completed could be baked in instead of having to be accounted for by each individual implementer.

I'm currently writing a simple shell script that will either be hooked up to a cron-job, or just be the entrypoint for the Docker container, that will do that logic.

I suppose the biggest issue I can see is that there are multiple use cases that the Daemon/my script should account for, which is evident based on your own script above:

1. Trigger sync of local file system change (your approach, only works if you are syncing local <-> remote)
2. Time based sync (my approach as I'm syncing remote <-> remote)
3. Other use cases I haven't yet considered.

**Comment by untitaker** [comment]
You can write a shell script that alternates between `vdirsyncer sync` and `sleep` and run that as a daemon. No need to check for overlapping executions then.

**Comment by untitaker** [comment]
FWIW I don't think I want this in vdirsyncer 1.0, but as a separate (unstable!) project for now.

**Comment by Decipher** [comment]
Yes, you are right, as if it is actively running `vdirsyncer sync` it won't continue the shell script until said process has finished. Clearly I was just overthinking it ;)

I still think that a daemon mode is something that would benefit the project, but understand the desire to not overcomplicate the the 1.0 release.

I do think putting it in a seperate project isn't the best approach though, as it limits the visibility of the feature. Maybe it doesn't even need to be a daemon per se, just a flag for `sync` that keeps it actively running until it's exited or errors.

Anyway, I'll continue with the script approach for the moment and will let you know once the docker project is in a usable state.

**Comment by hobarrera** [comment]
Another bonus feature for daemon mode is "ignore one-time network errors" (or maybe just warn). For example, if the network connection fails *once* in a row ignore it, and only warn me on the second attempt.

I occasionally get cron emails due to a transient network error that's not there any more 5 minutes later. If I disable cron emails, I'll never find about about conflicts.

**Comment by untitaker** [comment]
The thing is, all of this (even the retry logic) can be implemented in a wrapper script. I don't really see the need for this to be in vdirsyncer anymore.

**Comment by WhyNotHugo** [comment]
On the topic of RFC 6578, it seems that Xandikos supports that too.

As for a daemon mode, I'm using [`autovdirsyncer`](https://git.sr.ht/~whynothugo/autovdirsyncer), which is very basic but works. A `sync` is run automatically when a local file changes, and otherwise a sync happens every 15 minutes.

The remaining issue here is that events created on the server take up to 15 minutes to get pulled, and polling isn't the most efficient thing. JMAP addresses this, but we don't support JMAP yet (#266). [This draft spec](https://datatracker.ietf.org/doc/html/draft-gajda-dav-push-00#section-1) exists too, but doesn't really work for us since (a) apparently no-one supports it (b) systems running `vdirsyncer` are seldom reachable remotely.
